Context loss recovery

nadako opened the pull request:
See commit messages for more info

nadako committed:
Make maskShader non-static as we want fresh mask Shader instance for a new GLMaskManager instance.

Unfortunately we have to cast from AbstractMaskManager to GLMaskManager to access it now,
but that's a general issue with the current RenderSession/Renderer mess and we'll look into it separately.

nadako committed:
Force dirty for DisplayObject's cache Bitmap when calling __forceRenderDirty on it

nadako committed:
Do not call onWindowCreate (a.k.a. create renderer) twice

Not much to comment here, it was called twice because of callback-hell and custom OpenFL Window implementation.
The hell is still there but at least the renderer is created only once.

nadako committed:
clear context3D object on render context loss

nadako committed:
force display tree to re-render itself on render context restore

this will invalidate batch quads and other stuff

nadako committed:
reset compressed texture formats support state on render context restore

because to re-enable the support we need to call getExtension again

Neu2o commented on src/openfl/_internal/renderer/opengl/GLMaskManager.hx:
```diff
@@ -24,7 +24,7 @@ import openfl.utils.ByteArray;
 	
 	
-	@:allow(openfl._internal.renderer.opengl) private static var maskShader = new GLMaskShader ();
+	public var maskShader(default,null) = new GLMaskShader ();
```
Was it needed to recreate the mask shader on context loss, while otherwise as a `static` it would just exist once?

nadako replied:
Yes, because `Shader` instances reference the gl context once initialized. We could un-init it, but I think simply having a new instance is cleaner. Also this is what's done to the default shader.

Neu2o commented on src/openfl/display/Stage.hx:
```diff
@@ -374,12 +375,6 @@ class Stage extends DisplayObjectContainer implements IModule {
-			onWindowCreate (window);
-			
-		}
-		
```
This is to avoid creating the renderer twice right?

nadako replied:
Yes, as mentioned in the relevant commit message :)

nadako committed:
add isBuffer/isTexture checks in `dispose` methods so it's safe to call them after context loss

since these are not called often, there should be no real performance hit because of this

nadako reviewed (commented):
interestingly, i don't get warnings from `GLTextureBase.dispose`, which might mean that starling throws away the object without disposing it, in case of context recovery. will investigate further!

Neu2o commented on src/openfl/display/Stage.hx:
```diff
@@ -682,13 +677,20 @@ class Stage extends DisplayObjectContainer implements IModule {
+		if (stage3D.context3D != null) {
+			stage3D.__contextRequested = true; // because we want to dispatch `context3DCreate`
+		}
+		stage3D.context3D = null;
```
I think I prefer to have something like a `__reset` function in `Stage3D` instead of setting it directly. And because `resize` or `renderStage3D` are always iterating all 3D stages, I think we should do it here as well:

```
for (stage3D in stage.stage3Ds) {
	
	stage3D.__reset ();
	
}
```

Something like this?

nadako replied:
It is (as usual) weirdly inconsistent in OpenFL in general. E.g. if we look into `GLRenderer` ctor, we'll see that it also only accesses `stage.stage3Ds[0]` directly, so I don't know... As for the reset function, it surely makes sense. Though then again, maybe we should do a similar refactoring in the `GLRenderer` constructor...

nadako replied:
Changed in d8d88958a35e015ceaef46c02311a41722656ec4

Neu2o commented on src/openfl/display/Stage.hx:
```diff
@@ -682,13 +677,20 @@ class Stage extends DisplayObjectContainer implements IModule {
 	
 	public function onRenderContextRestored (renderer:Renderer, context:RenderContext):Void {
 		
+		GLTextureBase.reset();
```
Maybe move this to `onRenderContextLost` to have all the resetting done on one function?

nadako replied:
It was there originally, but then there was a case with texture formats being requested between lost and restore handlers, breaking it again. I believe it was happening when async loading and texture creation (and thus supported format query) happens when the context is lost. So this way it's a bit less fragile.

nadako committed:
encapsulate Stage3D context lose behaviour (review remark)

nadako committed:
code style :)
